src/Eloquent/HasSchemaVersion.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+declare(strict_types=1);
+
+namespace MongoDB\Laravel\Eloquent;
+
+trait HasSchemaVersion
+{
+    public $currentSchemaVersion = 1;
+
+    /**
+     * Auto call on model instance as booting
+     *
+     * @return void
+     */
+    public static function bootHasSchemaVersion(): void
+    {
+        static::saving(function ($model) {
+          if (!$model->getSchemaVersion()) {
+              $model->setSchemaVersion($model->currentSchemaVersion);
+          }
+        });
+
+        static::retrieved(function ($model) {
+            $model->upgradeSchemaVersion();
+        });
+    }
+
+    /**
+     * Migrate document to the current schema version.
+     */
+    public function migrateSchema(int $fromVersion): void
+    {
+    }
+
+    /**
+     * migrate schema and set current model version
+     * @return void
+     */
+    public function upgradeSchemaVersion(): void
```

Comment:
Why do you need this public method? All this code can be safely in the "retrieved" closure.